aptos-move/framework/aptos-framework/sources/account/common_account_abstractions/ethereum_derivable_account.move

```diff
@@ -0,0 +1,382 @@
+    fun deserialize_abstract_public_key(abstract_public_key: &vector<u8>): SIWEAbstractPublicKey {
+        let stream = bcs_stream::new(*abstract_public_key);
+        let ethereum_address = *bcs_stream::deserialize_string(&mut stream).bytes();
+        let domain = *bcs_stream::deserialize_string(&mut stream).bytes();
+        SIWEAbstractPublicKey { ethereum_address, domain }
+    }
+
+    /// Returns a tuple of the signature type and the signature.
+    /// We include the issued_at in the signature as it is a required field in the SIWE standard.
+    fun deserialize_abstract_signature(abstract_signature: &vector<u8>): SIWEAbstractSignature {
+        let stream = bcs_stream::new(*abstract_signature);
+        let signature_type = bcs_stream::deserialize_u8(&mut stream);
+        if (signature_type == 0x00) {
+            let issued_at = bcs_stream::deserialize_string(&mut stream);
+            let signature = bcs_stream::deserialize_vector<u8>(&mut stream, |x| deserialize_u8(x));
+            SIWEAbstractSignature::EIP1193DerivedSignature { issued_at, signature }
+        } else {
+            abort(EINVALID_SIGNATURE_TYPE)
+        }
+    }
+
+    // construct a message that is used to verify the signature following the SIWE standard
+    // and ethers.js. ethers adds a prefix to the message, so we need to include it also
+    fun construct_message(
+        ethereum_address: &vector<u8>,
+        domain: &vector<u8>,
+        entry_function_name: &vector<u8>,
+        digest_utf8: &vector<u8>,
+        issued_at: &vector<u8>,
+    ): vector<u8> {
+        let message = &mut vector[];
+        message.append(*domain);
+        message.append(b" wants you to sign in with your Ethereum account:\n");
+        message.append(*ethereum_address);
+        message.append(b"\n\nTo execute transaction ");
+        message.append(*entry_function_name);
+        message.append(b" on Aptos blockchain");
+        let network_name = network_name();
+        message.append(b" (");
+        message.append(network_name);
+        message.append(b")");
+        message.append(b".");
+        message.append(b"\n\nURI: ");
+        message.append(*domain);
+        message.append(b"\nVersion: 1");
+        message.append(b"\nChain ID: ");
+        message.append(*string_utils::to_string(&chain_id::get()).bytes());
+        message.append(b"\nNonce: ");
+        message.append(*digest_utf8);
+        message.append(b"\nIssued At: ");
+        message.append(*issued_at);
+
+        let msg_len = vector::length(message);
+
+        let prefix = b"\x19Ethereum Signed Message:\n";
+        let msg_len_string = string_utils::to_string(&msg_len); // returns string
+        let msg_len_bytes = msg_len_string.bytes(); // vector<u8>
+
+        let full_message = &mut vector[];
+        full_message.append(prefix);
+        full_message.append(*msg_len_bytes);
+        full_message.append(*message);
+
+        aptos_hash::keccak256(*full_message)
```

Comment:
nit: I prefer if `construct_message` returns the whole message instead of the keccak256. It's much clearer in tests what the in/out should be. I would prefer to apply the keccak256 in authenticate_auth_data